### PR TITLE
Update the spec description of numeric types

### PR DESCRIPTION
**Pull request by mppf**
This PR adds information to the description of numeric types in the types.rst spec chapter. It adds examples and links to later sections since it forms a jumping-off point and it's fairly early in the spec, in terms of chapters, for anyone reading it consecutively.

One of the main motivations of this PR is to describe in the spec the behavior of casting between `real` and `imag` types.

I created this PR to address https://chapel.discourse.group/t/getting-the-imaginary-part-of-an-imaginary-number/31200

Reviewed by @vasslitvinov - thanks!

- [x] full comm=none testing

### CODE REVIEW

File: doc/rst/language/spec/types.rst [hunk 1]
```diff
@@ -147,6 +192,40 @@ Some statements require expressions of ``bool`` type and Chapel supports
...
+   logical negation of it:
+
+   .. code-block:: chapel
+
+      var x: bool = true;
+      var y = !x;
+      var z: bool;
+
+   All three variables have type ``bool``.  Note that the types of ``x``
+   an ``y`` are optional; the program indicates the type of ``x`` but
```
Comment (cassella) [inline doc/rst/language/spec/types.rst:211]: *and

File: doc/rst/language/spec/types.rst [hunk 2]
```diff
@@ -147,6 +192,40 @@ Some statements require expressions of ``bool`` type and Chapel supports
...
+   infers the type of ``y``.  See :ref:`Chapter-Variables` for more
```
Comment (cassella) [inline doc/rst/language/spec/types.rst:212]: I'd say it's the compiler that does the inferring, not the program.

File: doc/rst/language/spec/types.rst [hunk 3]
```diff
@@ -157,6 +236,9 @@ represent them. Valid bit-sizes are 8, 16, 32, and 64. The default
 signed integral type, ``int``, is a synonym for ``int(64)``; and the
 default unsigned integral type, ``uint``, is a synonym for ``uint(64)``.
 
+Variables of integral have a default value of ``0`` if they are
```
Comment (cassella) [inline doc/rst/language/spec/types.rst:239]: *"integral type"?

File: doc/rst/language/spec/types.rst [hunk 4]
```diff
@@ -175,30 +257,136 @@ uint(64), uint 0                    18446744073709551615
...
+
+ * in decimal without an exponent, e.g. ``5.2``
+ * in decimal with an exponent, e.g. ``6.02e23``
+ * in hexadecimal without an exponent, e.g. ``0x2.fe``
+ * in hexadecimal with a decimal exponent, e.g. ``0x2.fep23``
+
+   *Example (harmonic.chpl)*.
+
+   For example, this program computes the first ``n`` terms of harmonic
+   series.
```
Comment (cassella) [inline doc/rst/language/spec/types.rst:343]: *"of the harmonic series"

File: doc/rst/language/spec/types.rst [hunk 5]
```diff
@@ -133,6 +155,29 @@ representation at run-time.
...
+      }
+      g(1);    // outputs 1
+      g(none); // does not create output
+
+   .. BLOCK-test-chapeloutput
+
+       1
+
+
+
```
Comment (vasslitvinov) [inline doc/rst/language/spec/types.rst:180]: Fewer empty lines above?

File: doc/rst/language/spec/types.rst [hunk 6]
```diff
@@ -147,6 +192,41 @@ Some statements require expressions of ``bool`` type and Chapel supports
...
+      var x: bool = true;
+      var y = !x;
+      var z: bool;
+
+   All three variables have type ``bool``.  Note that the types of ``x``
+   and ``y`` are optional; the program indicates the type of ``x`` but
+   the compiler infers the type of ``y``.  See :ref:`Chapter-Variables`
+   for more details. The last variable is initialized to the default
+   value of ``bool``, which is ``false`` (see
+   :ref:`Default_Values_For_Types`).
```
Comment (vasslitvinov) [inline doc/rst/language/spec/types.rst:215]: I suggest merging the above two lines into one, for better looks of the rst.

File: doc/rst/language/spec/types.rst [hunk 7]
```diff
@@ -175,30 +258,136 @@ uint(64), uint 0                    18446744073709551615
...
 can implicitly convert to other numeric types that can losslessly store
 the value. See :ref:`Implicit_Compile_Time_Constant_Conversions`.
+Integer literals can be written in hexadecimal, octal, or binary. See
+:ref:`Literals`.
+
+Signed or unsigned integral types of smaller width can implicitly convert
+to signed integral types of larger width.  Additionally, unsigned
+integral types of a smaller width can implicitly convert to an unsigned
+integral type of larger width. See :`Implicit_NumBool_Conversions` for
+details.
```
Comment (vasslitvinov) [inline doc/rst/language/spec/types.rst:268]: It was a bit hard for me to read. How about this:

Signed integral types of can implicitly convert to signed integral types of larger width.  Unsigned integral types can implicitly convert to both signed and unsigned integral type of larger width. See ....

Also, should we say that int(64) can implicitly convert to uint(64) (or in the other direction), if that is the case?

File: doc/rst/language/spec/types.rst [hunk 8]
```diff
@@ -175,30 +258,136 @@ uint(64), uint 0                    18446744073709551615
...
+
+   .. code-block:: chapel
+
+      for i in 1..n {
+        sum += 1.0/i;
+      }
+
+   Note that it uses `1.0/i` in order to do a floating point division. If
+   it used `1/i`, it would do integer division (rounding towards zero),
+   which evaluates to 0 in this case.
```
Comment (vasslitvinov) [inline doc/rst/language/spec/types.rst:371]: "in this case" --> "for i > 1"

File: doc/rst/language/spec/types.rst [hunk 9]
```diff
@@ -209,6 +398,56 @@ include ``imag(32)`` and ``imag(64)``.
...
+not initialized to something else (see also :ref:`Chapter-Variables`).
+
+It is possible to convert between a ``real`` value and an ``imag`` value
+using an explicit cast (see :ref:`Explicit_Conversions`). Similarly, an
+``imag`` value can be cast to a ``real`` value. Such casts preserve the
+floating-point value while changing whether or not it is imaginary.
+
+   *Example (imaginary.chpl)*.
+
+   For example, this program creates a imaginary numbers in two different
```
Comment (vasslitvinov) [inline doc/rst/language/spec/types.rst:419]: "a imaginary" --> "an imaginary"

**Review by vasslitvinov — APPROVED** [review]
See my comments.

**Pull request merged**